### ISSUE TITLE
enhancement: "ipfs get" using the Filestore

### ISSUE THREAD

**Issue by ProgVal** [ipfs/go-ipfs#3981]
#### Type:

Enhancement

#### Severity:

Medium

#### Description:

`ipfs add` now has an experimental `--no-copy` option, which allows to make a file/directory known to IPFS, and serve it to other peers from its original place on the hard disk.

Could you add a similar feature to `ipfs get` to write files to the filesystem, without adding them in `~/.ipfs` too (assuming they are not already there)?

Thanks!

**Comment by whyrusleeping** [comment]
This should be doable, though it will be difficult. We could potentially do this by adding some sort of signaling method to the filestore that says "when you get a block by this hash, you should add it to the filestore with this filepath and offset".

Thoughts @kevina ?

**Comment by kevina** [comment]
Bypassing the "cache" would be difficult.  Copying something in the cache to disk without rehashing is fairly straightforward.  At least it was on my implementation as I implemented as I already implemented it https://github.com/ipfs-filestore/go-ipfs/blob/master/filestore/util/move.go.  Since the current implementation differs from mine i am not 100% sure how useful my code is.

**Comment by whyrusleeping** [comment]
@kevina the difference is that its not being moved to the disk after the fact, we want to write it to the file as its being fetch, avoiding the blockstore entirely

**Comment by ProgVal** [comment]
@whyrusleeping What I had in mind was to write in the blockstore until the file is complete, and then write to the file without recomputing its hash and clear the blockstore.
Your idea is a great feature, but as you said, it is much harder to implement.

**Comment by whyrusleeping** [comment]
@ProgVal ah, that could be done with first fetching the file with `ipfs get`, then running a gc (to clear the blocks from your blockstore) then re-adding it with a no-copy add. Of course, that could all be automated in a flag on get.

The problem there though is that running that gc has the unintended side effect of cleaning out the entire blockstore, which you don't necessarily want. removing blocks from the blockstore after the download is equally as difficult (Requires computing the entire pinset to make sure the blocks we want to remove arent pinned by other things already). I'm actually thinking that the best (and though maybe not the easiest, but the least problematic) option might actually be to try and stream from bitswap straight into the filestore (bypassing the blockstore as i suggest above).

**Comment by kevina** [comment]
@whyrusleeping there is the `ipfs block rm` command that is already implemented, you can batch remove blocks so the pinset is only computed once.

**Comment by kevina** [comment]
`ipfs filestore mv` would be a first step in doing the (better) way you described.  Getting a block from the network and not storing in the blockstore would be a next step.

**Comment by whyrusleeping** [comment]
I think doing `ipfs filestore mv` should be the first actionable on this issue.

**Comment by kevina** [comment]
@whyrusleeping `ipfs filestore mv` should be fairly easy, I created a new issue for it #4193

**Comment by whyrusleeping** [comment]
Great! I think that will be an exciting feature

**Comment by schomatis** [comment]
@kevina While you are moving forward with `ipfs filestore mv`, is there something I can contribute towards the `ipfs get --no-copy`?

**Comment by ildar** [comment]
Is the work on this stalled yet?

**Comment by Stebalien** [comment]
It looks like there was a start in #4261 but we'd also need something to then _remove_ the file from the blockstore.

**Comment by ildar** [comment]
One thing keeps rolling in my mind: if the remote file wasn't "raw-leaves"
then will filestore be able to provide the identical hashes?

**Comment by Stebalien** [comment]
Yes? But how would one do that?

**Comment by jl452** [comment]
"ipfs add" can use --chunker or --raw-leaves or not
...and after this hash will be sort of random
what hash i have after "ipfs get --no-copy" ?

1) can i take the same hash as the person who added the file?


ps:
i see hash-issue in ipfs, but it "official documented" https://github.com/ipfs/go-ipfs/pull/6891 from https://github.com/ipfs/go-ipfs/issues/4318
"Finally, a note on hash determinism. While not guaranteed, adding the same
file/directory with the same flags will almost always result in the same output
hash. However, almost all of the flags provided by this command (other than pin,
only-hash, and progress/status related flags) will change the final hash."

2) is there any type of hash in ipfs that contains several variants of hashes leading to the same file?

**Comment by Artoria2e5** [comment]
Another thing this can provide is a --pin option on `ipfs get`, making the help-with-seeding-but-also-look-at-files job somewhat more similar to other P2P applications.

**Comment by markg85** [comment]
Just a little "me too" reply, but with a reason why i want this feature.
Granted, https://github.com/ipfs/kubo/issues/8201 would be ideal but that seems to be long term.

In my usecase i'm syncing data between two IPFS nodes (using [syncthing](https://syncthing.net/)). This data is large per file (up till tens of gigabytes per file) which i only want to have once on my hdd.
Doing `ipfs get <cid> -o <whatever>` gives you a blockstore copy in the IPFS cache.
To fix that one would have to do:

```
ipfs repo gc
ipfs add --nocopy <file>
```

While that solution works, it's not sane because you have to `gc` your IPFS repo and then re-add that file.
Being able to do `ipfs get <cid> --nocopy -o <somefile>` would solve this issue.

**Comment by capricornusx** [comment]
**Proposal: one-pass UnixFS materialization (boxo assembler + opt-in Kubo command)**

`ipfs get` and `ipfs add --nocopy` are independent. Today get fetches every block into the blockstore, then tar-exports to disk. There is no way to write verified leaf bytes straight into the target layout and (when possible) register filestore refs in the same pass.

**Design**

1. **boxo:** a UnixFS assembler that walks the DAG using `filesize` / `blocksizes` / links (the same metadata `DagReader` / `dagmodifier` already use), preallocates files, then `GetBlocks` leaves and `WriteAt` out of order. Internal nodes stay in the regular blockstore; leaves are not copied into flatfs.
2. **kubo:** thin opt-in CLI/RPC on top of that (new flag or command). Default `ipfs get` unchanged. Gated on `Experimental.FilestoreEnabled` if filestore indexing is requested. Off-switch, docs, and CLI tests required.

This is the usual boxo-then-kubo split: the walker does not depend on Kubo config/daemon; the product surface stays in Kubo. Companion PRs.

**What this can and cannot do**

- Materialize **any** UnixFS CID according to its actual DAG. Chunking is not in the CID string; it is discovered by walking the tree.
- Stock filestore indexing (`CID → path + offset`) only works for **raw leaves**, same constraint as `add --nocopy`. Protobuf leaves can still be unwrapped to disk in one pass; serving those leaf CIDs from the file would need a wrap-on-read path that filestore does not have today. Out of scope for the first cut.
- `ipfs get -o` is a **client** path over RPC tar. Filestore lives on the **daemon** filesystem. Do not overload `-o`. Local-daemon CLI or a command that takes a daemon-local path.

**Out of scope**

- Changing the default CID recipe or default `get` behavior
- Removing filestore path sandbox (`FileManager` root)
- Stabilizing filestore / fixing #7161
- Incomplete-download UX beyond fail-and-clean, unless trivial

**Ask**

Is this scope acceptable as experimental? If yes, we would start with the boxo assembler + a companion Kubo draft PR, and keep filestore registration a separate follow-up if that helps review.

---
